File: native-views/react-native-native-list/android/src/main/java/com/onekey/nativelist/NativeListActionRenderer.kt
package com.margelo.nitro.nativelist

import android.view.Gravity
import com.facebook.react.uimanager.ThemedReactContext
import org.json.JSONArray
import org.json.JSONObject

internal class NativeListActionRowView(context: ThemedReactContext) :
  NativeListRendererRowView(context) {
  private val visual = NativeListLeadingVisual(context)
  private val title = NativeListTextView(context)
  private val accessories = NativeListAccessoryStack(context)
  override val assetFields = listOf("icon")

  init {
    visual.glyphSize = 24
    visual.iconBorder = false
    orientation = HORIZONTAL
    gravity = Gravity.CENTER_VERTICAL
    addView(visual)
    addView(title, LayoutParams(0, LayoutParams.WRAP_CONTENT, 1f))
    addView(accessories)
    accessories.onAction = { key, view, slot, target ->
      emitAction(key, view, "trailingAccessory", slot, target, accessories.anchorInset(view))
    }
  }

  private val titlePaintFlags = title.paintFlags
  private val fallbackPaintFlags = visual.fallbackTextView.paintFlags

  override fun defaultHeight(item: NativeListItem, layout: String) =
    when {
      item.json.optString("presentation") == "accountSelector" -> 48
      item.json.has("icon") -> 60
      else -> 44
    }

  override fun bindContent(
    item: NativeListItem,
    theme: JSONObject?,
    layout: String,
    checkboxState: (NativeListItem, NativeSelectionTarget?, String) -> String,
  ) {
    val style = item.json.optJSONObject("style") ?: JSONObject()
    val icon = item.json.optJSONObject("icon")
    val alignment = style.optJSONObject("container")?.optString("contentVerticalAlignment")
    gravity =
      when (alignment) {
        "top" -> Gravity.TOP
        "bottom" -> Gravity.BOTTOM
        else -> Gravity.CENTER_VERTICAL
      }
    val selector = item.json.optString("presentation") == "accountSelector"
    visual.roundedGlyphOrigin = selector && item.hasExplicitHeight
    val descriptors =
      item.json.optJSONObject("checkbox")?.let { JSONArray().put(it) }
        ?: item.json.optJSONArray("trailing")
        ?: JSONArray()
    accessories.bind(item, descriptors, theme, style, sourceScale, checkboxState)
    // Legacy: overhanging selector icons/checkboxes must not be clipped by row padding.
    clipToPadding = !accessories.requestsUnclippedHost
    val hp =
      if (style.has("horizontalPadding")) stylePx(style.optDouble("horizontalPadding"))
      else dp(if (layout == "table") 16 else 12)
    val vp =
      if (style.has("verticalPadding")) stylePx(style.optDouble("verticalPadding")) else dp(8)
    setPadding(hp, vp, if (style.has("horizontalPadding")) hp else accessories.endInset ?: hp, vp)
    val titleParams = title.layoutParams as LayoutParams
    // Legacy: the weighted title abuts the trailing accessories without a gap.
    titleParams.marginEnd = 0
    title.layoutParams = titleParams
    val tone = item.json.optString("tone")
    val token =
      if (selector && tone != "primary") "secondaryText"
      else if (!selector && tone == "danger") "negative" else "primaryText"
    val fallback =
      if (token == "negative") "#C40006D3"
      else if (selector || token == "secondaryText") "#0000009B" else "#000000DF"
    NativeListResolvedText.resolve(
        context,
        item.json.optString("title"),
        style.optJSONObject("title"),
        16f,
        if (selector && icon == null) "regular" else "medium",
        parseNativeListColor(theme?.optString(token, fallback) ?: fallback),
        24,
        1,
        sourceScale,
      )
      .bind(title)
    title.paintFlags = titlePaintFlags
    // Legacy selector typography covered every text in the row, including the
    // trailing accessory text; explicit style fontSize/lineHeight stay as set.
    if (selector && sourceScale) {
      applyNativeListSourceTypography(title, style.optJSONObject("title"))
      accessories.applySourceTypography(style)
    }
    visual.visibility = if (icon == null) GONE else VISIBLE
    // Legacy: the 1px icon outline is kept only for a non-selector icon with an
    // explicit background; selector icons use a plain 8dp rounded fill.
    visual.iconBorder = !selector && icon?.has("backgroundColor") == true
    if (icon != null) {
      val source = JSONObject(icon.toString())
      if (!source.has("backgroundColor"))
        source.put(
          "backgroundColor",
          if (selector) theme?.optString("strongBackground", "#0000000F") ?: "#0000000F"
          else "#00000000",
        )
      val image = JSONObject(style.optJSONObject("image")?.toString() ?: "{}")
      if (selector && !image.has("cornerRadius") && !image.has("shape"))
        image.put("cornerRadius", scaledDp(8f) / resources.displayMetrics.density)
      visual.layoutParams =
        LayoutParams(
            if (image.has("width")) stylePx(image.optDouble("width"))
            else dp(if (selector) 32 else 40),
            if (image.has("height")) stylePx(image.optDouble("height"))
            else dp(if (selector) 32 else 40),
          )
          .apply {
            marginEnd =
              if (style.has("leadingGap")) stylePx(style.optDouble("leadingGap"))
              // Legacy: Yoga rounds the cumulative selector edges, not each gap.
              else if (selector && item.hasExplicitHeight) dp(56) - dp(12) - dp(32)
              else dp(12)
          }
      visual.bind(source, image, item.key, theme, false, sourceScale)
    } else visual.recycle()
    visual.fallbackTextView.paintFlags = fallbackPaintFlags
    // Legacy selector typography also covered the leading fallback text.
    if (selector && sourceScale && icon != null)
      applyNativeListSourceTypography(visual.fallbackTextView, null)
  }

  override fun onLayout(changed: Boolean, l: Int, t: Int, r: Int, b: Int) {
    super.onLayout(changed, l, t, r, b)
    val item = tag as? NativeListItem ?: return
    if (
      item.hasExplicitHeight &&
        item.json.optString("presentation") == "accountSelector" &&
        item.json
          .optJSONObject("style")
          ?.optJSONObject("container")
          ?.has("contentVerticalAlignment") != true
    ) {
      for (view in listOf(title, accessories)) if (view.visibility != GONE)
        view.offsetTopAndBottom(
          paddingTop + (height - paddingTop - paddingBottom - view.height + 1) / 2 - view.top
        )
    }
  }

  override fun bindSelectionContent(
    item: NativeListItem,
    checkboxState: (NativeListItem, NativeSelectionTarget?, String) -> String,
  ) {
    accessories.updateSelection(item, checkboxState)
  }

  override fun recycleContent() {
    visual.recycle()
    accessories.reset()
  }

  override fun disposeContent() {
    visual.dispose()
    accessories.reset()
  }
}

File: native-views/react-native-native-list/ios/NativeListActionRenderer.swift
import UIKit

final class NativeListActionCell: NativeListRendererCell {
  private let visual = NativeListLeadingVisual(frame: .zero)
  private let title = NativeListTextLabel()
  private let accessories = NativeListAccessoryStack(frame: .zero)
  private var visualWidth: NSLayoutConstraint!
  private var visualHeight: NSLayoutConstraint!
  override var assetFields: [String] { ["icon"] }

  override init(frame: CGRect) {
    super.init(frame: frame)
    visual.glyphSize = 24
    root.axis = .horizontal
    root.alignment = .center
    [visual, title, accessories].forEach(root.addArrangedSubview)
    visual.translatesAutoresizingMaskIntoConstraints = false
    visualWidth = visual.widthAnchor.constraint(equalToConstant: 40)
    visualHeight = visual.heightAnchor.constraint(equalToConstant: 40)
    NSLayoutConstraint.activate([visualWidth, visualHeight])
    title.setContentHuggingPriority(UILayoutPriority(1), for: .horizontal)
    title.setContentCompressionResistancePriority(.defaultLow, for: .horizontal)
    accessories.onAction = { [weak self] key, view, slot, target in
      self?.emitAction(key, from: view, source: "trailingAccessory", slot: slot, target: target, anchorInset: self?.accessories.anchorInset(for: view) ?? 0)
    }
  }
  required init?(coder: NSCoder) { fatalError("init(coder:) has not been implemented") }
  override func bindContent(
    _ item: NativeListItem, theme: [String: Any]?, layout: String,
    checkboxState: (NativeListItem, NativeSelectionTarget?, String) -> String
  ) {
    let style = item.data.dictionary("style") ?? [:]
    root.alignment =
      style.dictionary("container")?.string("contentVerticalAlignment") == "top"
      ? .top
      : style.dictionary("container")?.string("contentVerticalAlignment") == "bottom"
        ? .bottom : .center
    let selector = item.data.string("presentation") == "accountSelector"
    let icon = item.data.dictionary("icon")
    let descriptors =
      item.data.dictionary("checkbox").map { [$0] } ?? item.data.dictionaries("trailing")
    accessories.bind(
      item, descriptors: descriptors, theme: theme, style: style, checkboxState: checkboxState)
    let hp = CGFloat(style.double("horizontalPadding", default: layout == "table" ? 16 : 12))
    let vp = CGFloat(style.double("verticalPadding", default: 8))
    let end =
      descriptors.count == 1 && descriptors[0].string("kind") == "chevron"
      ? 6 : accessories.endInset ?? hp
    contentInsets = UIEdgeInsets(
      top: vp, left: hp, bottom: vp, right: style["horizontalPadding"] == nil ? end : hp)
    root.spacing = 12
    root.setCustomSpacing(CGFloat(style.double("leadingGap", default: 12)), after: visual)
    root.setCustomSpacing(accessories.contentGap ?? 12, after: title)
    let tone = item.data.string("tone")
    let color =
      selector
      ? nativeListColor(
        theme, tone == "primary" ? "primaryText" : "secondaryText",
        tone == "primary" ? "#202020" : "#646464")
      : nativeListColor(
        theme, tone == "danger" ? "negative" : "primaryText",
        tone == "danger" ? "#CE2C31" : "#202020")
    NativeListResolvedText(
      item.data.string("title"), style: style.dictionary("title"), size: 16,
      weight: selector && icon == nil ? .regular : .medium, color: color, lineHeight: 24, lines: 1
    ).bind(title)
    visual.isHidden = icon == nil
    if var icon {
      if icon["backgroundColor"] == nil { icon["backgroundColor"] = "#00000000" }
      var image = style.dictionary("image") ?? [:]
      if selector && image["cornerRadius"] == nil && image["shape"] == nil {
        image["cornerRadius"] = 8
      }
      visualWidth.constant = CGFloat(image.double("width", default: selector ? 32 : 40))
      visualHeight.constant = CGFloat(image.double("height", default: selector ? 32 : 40))
      visual.bind(icon, style: image, key: item.key, theme: theme, isUnread: false)
      visual.layer.borderWidth = 0
    } else {
      visual.recycle()
    }
  }
  override func bindSelectionContent(
    _ item: NativeListItem,
    checkboxState: (NativeListItem, NativeSelectionTarget?, String) -> String
  ) {
    accessories.updateSelection(item, checkboxState: checkboxState)
  }
  override func recycleContent() {
    visual.recycle()
    accessories.reset()
  }
}
